Canonicalize triton::ReduceOp select(cmpf) patterns to min/max operations

quic-samanara opened the pull request:
This PR introduces a pattern rewrite that simplifies `triton::ReduceOp` bodies containing `arith.select` operations based on floating-point comparisons. The transformation replaces select-based logic with equivalent min/max operations, improving IR canonicalization and enabling further optimization. 
### Supported Transformations:
- `select(cmpf ogt a, b), a, b` → `arith.maxf(a, b)`
- `select(cmpf olt a, b), a, b` → `arith.minf(a, b)`
- `select((cmpf ogt a, b) || cmpf une a, a), a, b` → `arith.maximumf(a, b)`
- `select((cmpf olt a, b) || cmpf une a, a), a, b` → `arith.minimumf(a, b)`

Applying this transformation before the `ReduceOp` converter runs allows the natural lowering of `tt.reduce` to `linalg.reduce`; the motivational use case was from a torch-inductor spit out Triton kernel which consisted of this complex pattern in the reduction body and did not allow the `ReduceOp` to get lowered to the `linalg` dialect.

quic-samanara commented:
@microsoft-github-policy-service agree company="Qualcomm"

quic-samanara commented:
@nhat-nguyen, sorry for the explicit ping but I did not have access to request a formal review. Could you be pleaase review the PR / or request review from appropriate stakeholders? 
Thanks!

nhat-nguyen commented:
@quic-samanara thanks for the pr! it's been a while since i last looked at this pass, but there's a somewhat similar pattern: `MinMaxConverter` in `ConversionPatterns.hpp`. could we merge these somehow?

quic-samanara commented:
> @quic-samanara thanks for the pr! it's been a while since i last looked at this pass, but there's a somewhat similar pattern: `MinMaxConverter` in `ConversionPatterns.hpp`. could we merge these somehow?

@nhat-nguyen, thanks for the quick review. I tried to leverage the `MinMaxConverter` infrastructure and perform the rewrite under the required constraints there. Let me know how that looks. Thanks!

nhat-nguyen commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1871,6 +1911,112 @@ struct MinMaxConverter : public OpRewritePattern<CmpOp> {
+  /// operations, which is already supported in the triton-shared dialect
+  /// conversion pipeline.
+
+  LogicalResult foldCmpSelectToMinMax(PatternRewriter &rewriter,
```
i think this function has a lot of similarities with the existing code in this pattern. could we try to consolidate them?

nhat-nguyen commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1856,7 +1880,23 @@ struct MinMaxConverter : public OpRewritePattern<CmpOp> {
+      selectOp = findSelectOpConsumingCmp(cmpOp.getResult());
+      if (!selectOp) {
+        return failure();
+      }
```
The first condition in `findSelectOpConsumingCmp` is already being handled by line `1882`. So in reality we would only be hitting the `OrOp` case. Can we refactor this?

nhat-nguyen commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1856,7 +1880,23 @@ struct MinMaxConverter : public OpRewritePattern<CmpOp> {
+      // cmpOp. We can proceed to check if the select op is part of the same
+      // reduce op.
+      auto reduceOp = findEnclosingReduceOp(selectOp);
+      if (reduceOp != findEnclosingReduceOp(cmpOp)) {
```
this pattern doesn't require us to be inside a reduce op, so i think we can relax this condition

quic-samanara committed:
Refine MinMaxConverter

Address review comments to reuse code

quic-samanara commented:
Hi @nhat-nguyen, addressed some of the comments you had:

- Introduced a `foldCmpToMinMax` helper that centralizes the mapping from `arith.cmpf` predicates to the appropriate Min/Max operations, and reuse it from the generic `cmp+select` pattern and the NaN-aware folding logic. 
- Make the `cmp->OrIOp->select` folding logic operate directly on `arith.select` relaxing the `triton::ReduceOp` restriction.
- Simplify the control flow in `matchAndRewrite` by 
    * The direct cmp->select case is handled "inline". 
    * Restricting redundant checking to fire the complex NaN-aware fold pattern. 

Hope this addresses some of your concerns.

quic-samanara commented:
Hi, pinging here to check if more changes are needed. Thanks!

nhat-nguyen commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1845,47 +1845,205 @@ template <typename CmpOp>
+        if (pred == arith::CmpFPredicate::OGT ||
+            pred == arith::CmpFPredicate::OLT) {
+          rewriter.setInsertionPoint(sel);
+          Value foldedResult;
```
nit: could you add an insertion guard here just in case we accidentally use the incorrect insertion point later? same for other places too

nhat-nguyen commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1845,47 +1845,205 @@ template <typename CmpOp>
+  /// numeric operations (MaxNumFOp/MinNumFOp).
+  LogicalResult foldCmpToMinMax(PatternRewriter &rewriter, Location loc,
+                                Value lhs, Value rhs, arith::CmpFPredicate pred,
+                                bool useNaNOps, Value &result) const {
```
i think you can use `FailureOr<Value>` to avoid having to pass a reference storing the result :)

quic-samanara replied:
Thanks for the suggestion, done.

nhat-nguyen commented:
@quic-samanara looks good to me, just left a few minor comments. i don't have permission to run the workflow or approve anymore though. pinging @bmyerz0 @red1bluelost for help

quic-samanara committed:
Use Failure<Or> insteaf of passing in reference
to output parameter.

quic-samanara committed:
Add insertion guards to ensure that the insertion
point is restored after pattern rewrites.

quic-samanara commented:
@nhat-nguyen, thanks a lot for the review. Addressed the last two comments you pointed.

red1bluelost commented:
Could you add a lit test for this.

red1bluelost commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1845,49 +1845,209 @@ template <typename CmpOp>
 struct MinMaxConverter : public OpRewritePattern<CmpOp> {
```
This whole converter should probably be a pattern for the select operation and not the compare operation. I'm not confident it is best practice, if not plain illegal, to replace an operation that is not the root of the pattern.

No action necessary for this PR since your work is separate from this root issue. I'm just mentioning the concern here since this review is in the same area which led me to notice the issue.

red1bluelost commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1845,49 +1845,209 @@ template <typename CmpOp>
+  ///
+  /// 1. ** Simple Min/Max Reduction **
+  ///    - select (cmpf ogt a, b), a, b --> arith.maxnumf(a, b)
+  ///    - select (cmpf olt a, b), a, b --> arith.minnumf(a, b)
```
From what I can tell `foldCmpSelectToMinMax does not perform this pattern because it would be handled by the if statement at line 1903.

red1bluelost commented on include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp:
```diff
@@ -1845,49 +1845,209 @@ template <typename CmpOp>
+    Value falseVal = sel.getFalseValue();
+
+    // Case 1: Simple Min/Max Reduction (numeric min/max).
+    if (auto cmp = dyn_cast<arith::CmpFOp>(condOp)) {
```
As mentioned in other comment, this case would be unreachable.

quic-samanara replied:
Thanks for catching that, changed accordingly.

quic-samanara committed:
Simplify control flow in foldCmpSelectToMinMax
pattern

quic-samanara committed:
Add LIT test

quic-samanara commented:
> Could you add a lit test for this.

Thanks for reviewing @red1bluelost, addressed your comments and added a LIT test.

red1bluelost commented:
No more PRs are being accepted on this repo anymore. Please see https://github.com/microsoft/triton-shared/pull/367

quic-samanara commented:
> No more PRs are being accepted on this repo anymore. Please see #367

@red1bluelost, oh, I see. Does that mean there is going to be no development from now on? Is someone else taking the ownership?
